File: lua/lifegame1.lua
-- KISS conway's game of life impl ---------------------------

local randomseed, rand, floor, abs = math.randomseed, math.random, math.floor, math.abs
local band, rsh, lsh = bit.band, bit.rshift, bit.lshift
local random = function(n) 
  n = n or 1; 
  return floor(rand()*abs(n)) 
end

function bench(desc, f)
  local t = os.clock()
  f()
  print( desc or "", os.clock() - t )
end

-------------

local function new_grid(w, h) 
  local grid = {}
  w, h = w or 15, h or 15
  for y = 1, h+2 do
    grid[y] = {}
    for x = 1, w+2 do
      grid[y][x] = 0
    end
  end
  return grid
end

local function grid_print(grid, w, h)
  w, h = w or 15, h or 15
  if not grid then return end
  for y = 2, h+1 do
    for x = 2, w+1 do 
      io.write(string.format("%d ", grid[y][x]))
    end
    print()
  end
end

local function neighbor_count(old, y, x, h, w)
  --[[local count = 0
  for i=1, #dir do
    local ny, nx = y + dir[i][1], x + dir[i][2]
    if ny < 1 then ny = h
    elseif ny > h then ny = 1 end
    if nx < 1 then nx = w
    elseif nx > w then nx = 1 end
    if old_grid[ny][nx] > 0 then count = count + 1 end
  end]]
  local count = (old[y-1][x-1] + old[y-1][x] + old[y-1][x+1]) +
                (old[y][x-1]   +               old[y][x+1])   +
                (old[y+1][x-1] + old[y+1][x] + old[y+1][x+1])
  return count
end

--local rule1 = ffi.new("char[9]", {0, 0, 1, 1, 0, 0, 0, 0, 0})
--local rule2 = ffi.new("char[9]", {0, 0, 0, 1, 0, 0, 0, 0, 0})
--what if we need different rule sets? how to do that with bitwise trick??
local function ruleset(now, count)
  return band(rsh(lsh(now, 2) + 8, count), 1)
end

local function wrap_padding(old, w, h)
  w, h = w or 15, h or 15
  -- side wrapping.
  for x = 3, w do 
    old[h+2][x] = old[ 2 ][x]
    old[ 1 ][x] = old[h+1][x]
  end
  for y = 3, h do 
    old[y][w+2] = old[y][ 2 ]
    old[y][ 1 ] = old[y][w+1]
  end
  -- .. and corner wrapping, obviously I am too stupid.
  local tl, tr, bl, br = old[2][2], old[2][w+1], old[h+1][2], old[h+1][w+1]
  old[2][w+2], old[h+2][2], old[h+2][w+2] = tl, tl, tl
  old[2][1],   old[h+2][1], old[h+2][w+1] = tr, tr, tr
  old[1][2],   old[1][w+2], old[h+1][w+2] = bl, bl, bl
  old[1][1],   old[1][w+1], old[h+1][1]   = br, br, br
end

local function grid_iteration(old, new, w, h)
  w, h = w or 15, h or 15
  wrap_padding(old)
  for y = 2, h+1 do
    for x = 2, w+1 do
      new[y][x] = ruleset( old[y][x], neighbor_count(old, y, x, h, w) )
    end
  end
  -- new and old can be used interchangably, no need to copy here
end

---------

local now = new_grid(20, 20)
local new = new_grid(20, 20)
randomseed(os.time())

local function test_by_hand()
  for i=1, 80 do
    now[random(20)+2][random(20)+2] = 1  -- random seeding
  end
  local i, index = 0, 0
  local grids = {}
  grids[0], grids[1] = now, new
  while true do
    index = i % 2
    grid_iteration( grids[index], grids[bit.bxor(index, 1)], 20, 20 )
    grid_print( grids[bit.bxor(index, 1)], 20, 20)
    io.read()
    i = i + 1
  end
end

local function bench_test(n)
  for i=1, 80 do
    now[random(20)+2][random(20)+2] = 1  -- random seeding
  end
  local function performance_test(n, now, new)
    print("Memory usage before first run: "..collectgarbage("count").." KiB.")
    local index = 0
    local grids = {}
    grids[0], grids[1] = now, new
    for i = 0, n-1 do
      index = i % 2
      grid_iteration(grids[index], grids[bit.bxor(index, 1)], 20, 20)
    end
    print("Memory usage after last run: "..collectgarbage("count").." KiB.")
  end
  grid_print(now, 20, 20)
  bench(string.format("Conway's Game of Life %d iterations: ", n), 
        function() return performance_test(n, now, new) end)
  grid_print(now, 20, 20)
end

bench_test(100000)

--test_by_hand()

File: cpp/lifegame1.cpp
#include <cstdlib>
#include <cstdio>
#include <cmath>
#include <ctime>

template<typename T, size_t W, size_t H>
void grid_print(T const (&grid)[H][W])
{
    if( !grid || !H || !W ) return;
    for( size_t y = 0; y < H; ++y ) {
        for( size_t x = 0; x < W; ++x )
            printf("%d ", grid[y][x]);
        printf("\n");
    }
}

int dir[8][2] = {{-1,-1}, {-1,0}, {-1,1}, {0,-1}, {0,1}, {1,-1}, {1,0}, {1,1}};
template<typename T, size_t W, size_t H>
int neighbor_count(T const (&old_grid)[H][W], int y, int x)
{
    int count = 0;
    for( int i = 0; i < 8; ++i ) {
        int ny = y + dir[i][0];
        int nx = x + dir[i][1];
        if( ny < 0 ) ny = H-1;
        else if( static_cast<size_t>(ny) >= H ) ny = 0;
        if( nx < 0 ) nx = W-1;
        else if( static_cast<size_t>(nx) >= W ) nx = 0;
        if( old_grid[ny][nx] > 0 )
            count = count + 1;
    }
    return count;
}

int rule1[9] = {0, 0, 1, 1, 0, 0, 0, 0, 0};
int rule2[9] = {0, 0, 0, 1, 0, 0, 0, 0, 0};
int ruleset(int now, int n)
{
    return now > 0 ? rule1[n] : rule2[n];
}

template<typename T, size_t W, size_t H>
void grid_iteration(T (&old_grid)[H][W], T (&new_grid)[H][W])
{
    for( size_t y = 0; y < H; ++y )
        for( size_t x = 0; x < W; ++x )
            new_grid[y][x] = ruleset( old_grid[y][x], neighbor_count(old_grid, y, x) );

    for( size_t y = 0; y < H; ++y )
        for( size_t x = 0; x < W; ++x ) {
            old_grid[y][x] = new_grid[y][x]; // grid data copy
            new_grid[y][x] = 0;              // clean new grid data
        }
}

int nowg[15][15] = {{0}};
int newg[15][15] = {{0}};

void bench_test(int n = 0)
{
    for( int i = 0; i < 45; ++i )
        nowg[rand()%15][rand()%15] = 1;  // random seeding

    grid_print(nowg);
    time_t t = clock();
    for( int i = 0; i < n; ++i )
        grid_iteration(nowg, newg);
    printf("Time Elapsed: %lf\n", static_cast<double>(clock() - t) / CLOCKS_PER_SEC);
    grid_print(nowg);
}

int main()
{
    std::srand(time(0));
    bench_test(20000);
    system("pause");
    return 0;
}
